src/github.com/matrix-org/dendrite/publicroomsapi/storage/storage.go

```diff
@@ -0,0 +1,250 @@
+// If the event doesn't match with any property used to compute the public room directory,
+// does nothing.
+// If something went wrong during the process, returns an error.
+func (d *PublicRoomsServerDatabase) UpdateRoomFromEvent(
+	event gomatrixserverlib.Event, eventsToRemove []gomatrixserverlib.Event,
+) error {
+	roomID := event.RoomID()
+
+	// Process the event according to its type
+	switch event.Type() {
+	case "m.room.create":
+		return d.statements.insertNewRoom(roomID)
+	case "m.room.member":
+		return d.updateNumJoinedUsers(event, eventsToRemove, roomID)
+	case "m.room.aliases":
+		return d.updateRoomAliases(event, roomID)
+	case "m.room.canonical_alias":
+		var content common.CanonicalAliasContent
+		field := &(content.Alias)
+		attrName := "canonical_alias"
+		return d.updateStringAttribute(attrName, event, &content, field, roomID)
+	case "m.room.name":
+		var content common.NameContent
+		field := &(content.Name)
+		attrName := "name"
+		return d.updateStringAttribute(attrName, event, &content, field, roomID)
+	case "m.room.topic":
+		var content common.TopicContent
+		field := &(content.Topic)
+		attrName := "topic"
+		return d.updateStringAttribute(attrName, event, &content, field, roomID)
+	case "m.room.avatar":
+		var content common.AvatarContent
+		field := &(content.URL)
+		attrName := "avatar_url"
+		return d.updateStringAttribute(attrName, event, &content, field, roomID)
+	case "m.room.history_visibility":
+		var content common.HistoryVisibilityContent
+		field := &(content.HistoryVisibility)
+		attrName := "world_readable"
+		strForTrue := "world_readable"
+		return d.updateBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	case "m.room.guest_access":
+		var content common.GuestAccessContent
+		field := &(content.GuestAccess)
+		attrName := "guest_can_join"
+		strForTrue := "can_join"
+		return d.updateBooleanAttribute(attrName, event, &content, field, strForTrue, roomID)
+	}
+
+	// If the event type didn't match, return with no error
+	return nil
+}
+
+// updateNumJoinedUsers updates the number of joined user in the database representation
+// of the room identified by a given room ID, using a given Matrix event.
+// If the event's membership is "join", increments the value if the event isn't an update,
+// if not, decrements it.
+// Returns an error if the update failed.
+func (d *PublicRoomsServerDatabase) updateNumJoinedUsers(
+	membershipEvent gomatrixserverlib.Event, eventsToRemove []gomatrixserverlib.Event, roomID string,
+) error {
+	membership, err := membershipEvent.Membership()
+	if err != nil {
+		return err
+	}
+
+	if membership == "join" {
+		for _, event := range eventsToRemove {
+			if event.Type() == "m.room.member" {
+				// This is an update for an user who already joined the room, most
+				// likely triggered by a profile update. We don't increment the
+				// number of joined member in this case because we already counted
+				// this user.
+				return nil
```

Comment:
Does this work if there was more that one join event in the state added?
Would it be better to increment the count for each join in the state added and decrement the count for each join in the state removed?